rfcs/0006-page-transitions.md

```diff
@@ -0,0 +1,144 @@
+- Start Date: 2020-01-13
+- Project: [next-with-moxy](https://github.com/moxystudio/next-with-moxy)
+- Reference Issues: N/A
+- Implementation PR: N/A
+
+# Summary
+
+This proposal is to add full page transitions to the `next-with-moxy` boilerplate, either as a guide and easy-to-use feature (recipe) or an out-of-the-box feature (by including it in the source code).
+
+# Basic example
+
+We will use [Motion](https://www.framer.com/motion/) by Framer, an open-source and production-ready motion library for React on the web.
+
+## Installation
+
+First, install the package:
+
+```sh
+$ npm install --save framer-motion
+```
+
+## Component hierarchy
+
+To animate components as they're removed from the React tree, we need an [AnimatePresence](https://www.framer.com/api/motion/animate-presence/) component that wraps motion components (`Transition` in this case).
+
+In our [`App.js`](https://github.com/moxystudio/next-with-moxy/blob/master/www/app/App.js) file:
+
+```js
+render() {
+    const { Component, pageProps, router } = this.props;
+
+    return (
+        <>
+            <Head>
+                {/* ... */}
+            </Head>
+            <KeyboardOnlyOutlines>
+                <AnimatePresence exitBeforeEnter>
+                    <Transition key={ router.route }>
+                        <Component { ...pageProps } />
+                    </Transition>
+                </AnimatePresence>
+            </KeyboardOnlyOutlines>
+        </>
+    );
+}
+```
+
+## Motion components
+
+`Transition` is a [motion component](https://www.framer.com/api/motion/component/) that contains the animation to display when switching pages. A motion component works like their HTML counterparts, but offer some props that allow you to:
+
+- Declaratively or imperatively animate components.
+- Add drag, pan, hover and tap gestures.
+- Respond to gestures with animations.
+- Deeply animate throughout React trees via variants.
+
+A good place to put these would be in the [`www/shared/components`](https://github.com/moxystudio/next-with-moxy/tree/master/www/shared/components), in a *transitions* folder for example.
+
+```js
+import React from 'react';
+import PropTypes from 'prop-types';
+import { motion } from 'framer-motion';
```

Comment:
Is it really worth to include such a big lib just for page transitions?

Motion has a lot of code just for Drag and Spring physics, but do we really need them when transitioning between pages?

A simple css only solution would be better for the use-case represented in this example:

```css
@keyframes pageEnter {
    from {
         opacity: 0;
         transform: translateX(-600px)
    }
    to {
        opacity: 1; transform: translateX(0);
    }
}

@keyframes pageExit {
    from {
         opacity: 1;
         transform: translateY(0)
    }
    to {
        opacity: 0; transform: translateY(300px);
    }
}

.page.enter {
    animation: pageEnter 0.8s ease both;
}

.page.exit {
    animation: pageExit 0.5s ease forwards;
}
```


I know that you can argue that "it's simpler and faster to use js only with a custom lib". But is it really worth the extra bundle size?

Reply:
More data into my explanation:
https://bundlephobia.com/result?p=framer-motion@1.8.3

Reply:
I haven’t analyzed it, but perhaps the package is modular and we can import only what we need?